src/pandapipes/component_models/heat_consumer_component.py

```diff
@@ -119,50 +137,60 @@ def adaption_after_derivatives_hydraulic(cls, net, branch_pit, node_pit, idx_loo
-    #                                         - hs_pit[mask_t_return, DELTAT])
-    #
-    #
-    # @classmethod
-    # def adaption_after_derivatives_thermal(cls, net, branch_pit, node_pit, idx_lookups, options):
-    #     """
-    #
-    #     :param net:
-    #     :type net:
-    #     :param branch_component_pit:
-    #     :type branch_component_pit:
-    #     :param node_pit:
-    #     :type node_pit:
-    #     :return:
-    #     :rtype:
-    #     """
-    #     # -(rho * area * cp * v_init * (-t_init_i + t_init_i1 - tl)
-    #     #   - alpha * (t_amb - t_m) * length + qext)
-    #
-    #     f, t = idx_lookups[cls.table_name()]
-    #     hs_pit = branch_pit[f:t, :]
-    #     from_nodes = hs_pit[:, FROM_NODE_T].astype(np.int32)
-    #
-    #     mask_qext = ~np.isnan(hs_pit[:, QEXT])
-    #     mask_deltat = ~np.isnan(hs_pit[:, DELTAT])
-    #     mask_t_return = ~np.isnan(hs_pit[:, TRETURN])
-    #     mask_mass = ~np.isnan(hs_pit[:, MASS])
-    #     hs_pit[mask_t_return | mask_deltat, JAC_DERIV_DT1] = 0
+        consumer_array = get_component_array(net, cls.table_name())
+
+        hc_pit = branch_pit[f:t, :]
+        hc_pit[:, JAC_DERIV_DP] = 0
+        hc_pit[:, JAC_DERIV_DP1] = 0
+        hc_pit[:, JAC_DERIV_DM] = 1
+        hc_pit[:, LOAD_VEC_BRANCHES] = 0
+
+        mask = consumer_array[:, cls.MODE] == cls.QE_TR
+        if np.any(mask):
+            cp = get_branch_cp(net, get_fluid(net), node_pit, hc_pit)
+            from_nodes = hc_pit[mask, FROM_NODE_T].astype(int)
+            t_out = consumer_array[mask, cls.TRETURN]
+            t_mask = hc_pit[mask, TOUTINIT] == node_pit[from_nodes, TINIT]
+            node_pit[from_nodes[t_mask], TINIT] += 10
+            t_in = node_pit[from_nodes, TINIT]
+            df_dm = - cp[mask] * (t_out - t_in)
+            hc_pit[mask, LOAD_VEC_BRANCHES] = (- consumer_array[mask, cls.QEXT] + df_dm
+                                               * hc_pit[mask, MDOTINIT])
+            hc_pit[mask, JAC_DERIV_DM] = df_dm
+
+    @classmethod
+    def adaption_before_derivatives_thermal(cls, net, branch_pit, node_pit, idx_lookups, options):
+        f, t = idx_lookups[cls.table_name()]
+        hc_pit = branch_pit[f:t, :]
+        consumer_array = get_component_array(net, cls.table_name(), mode='heat_transfer')
+        mask = consumer_array[:, cls.MODE] == cls.MF_DT
+        if np.any(mask):
+            cp = get_branch_cp(net, get_fluid(net), node_pit, hc_pit)
+            q_ext = cp[mask] * consumer_array[mask, cls.MASS] * consumer_array[mask, cls.DELTAT]
+            hc_pit[mask, QEXT] = q_ext
+
+        mask = consumer_array[:, cls.MODE] == cls.MF_TR
+        if np.any(mask):
+            cp = get_branch_cp(net, get_fluid(net), node_pit, hc_pit)
+            from_nodes = hc_pit[mask, FROM_NODE_T].astype(int)
+            t_in = node_pit[from_nodes, TINIT]
+            t_out = hc_pit[mask, TOUTINIT]
+            q_ext = cp[mask] * consumer_array[mask, cls.MASS] * (t_in - t_out)
```

Comment:
We would assume that `consumer_array[mask, cls.MASS]` is the same as `hc_pit[mask, MDOTINIT]`, right? Do we need some assertion for this?